File: lib/features/support/domain/entities/support_request.dart
import 'package:equatable/equatable.dart';
import '../../../../l10n/app_localizations.dart';

enum SupportSubject {
  bug,
  question,
  suggestion,
  other;

  String get label {
    switch (this) {
      case SupportSubject.bug:
        return 'Signalement de bug';
      case SupportSubject.question:
        return 'Question';
      case SupportSubject.suggestion:
        return 'Suggestion';
      case SupportSubject.other:
        return 'Autre';
    }
  }

  String localizedLabel(AppLocalizations l10n) {
    switch (this) {
      case SupportSubject.bug:
        return l10n.support_subjectBug;
      case SupportSubject.question:
        return l10n.support_subjectQuestion;
      case SupportSubject.suggestion:
        return l10n.support_subjectSuggestion;
      case SupportSubject.other:
        return l10n.support_subjectOther;
    }
  }

  String get value {
    switch (this) {
      case SupportSubject.bug:
        return 'bug';
      case SupportSubject.question:
        return 'question';
      case SupportSubject.suggestion:
        return 'suggestion';
      case SupportSubject.other:
        return 'other';
    }
  }

  static SupportSubject fromString(String value) {
    switch (value) {
      case 'bug':
        return SupportSubject.bug;
      case 'question':
        return SupportSubject.question;
      case 'suggestion':
        return SupportSubject.suggestion;
      default:
        return SupportSubject.other;
    }
  }
}

enum SupportStatus {
  pending,
  inProgress,
  resolved,
  closed;

  String get label {
    switch (this) {
      case SupportStatus.pending:
        return 'En attente';
      case SupportStatus.inProgress:
        return 'En cours';
      case SupportStatus.resolved:
        return 'Résolu';
      case SupportStatus.closed:
        return 'Fermé';
    }
  }

  String localizedLabel(AppLocalizations l10n) {
    switch (this) {
      case SupportStatus.pending:
        return l10n.support_statusPending;
      case SupportStatus.inProgress:
        return l10n.support_statusInProgress;
      case SupportStatus.resolved:
        return l10n.support_statusResolved;
      case SupportStatus.closed:
        return l10n.support_statusClosed;
    }
  }

  static SupportStatus fromString(String value) {
    switch (value) {
      case 'in_progress':
        return SupportStatus.inProgress;
      case 'resolved':
        return SupportStatus.resolved;
      case 'closed':
        return SupportStatus.closed;
      default:
        return SupportStatus.pending;
    }
  }
}

class SupportRequest extends Equatable {
  final String? id;
  final SupportSubject subject;
  final String message;
  final SupportStatus status;
  final String? adminNotes;
  final DateTime? createdAt;
  final DateTime? resolvedAt;

  const SupportRequest({
    this.id,
    required this.subject,
    required this.message,
    this.status = SupportStatus.pending,
    this.adminNotes,
    this.createdAt,
    this.resolvedAt,
  });

  @override
  List<Object?> get props => [
        id,
        subject,
        message,
        status,
        adminNotes,
        createdAt,
        resolvedAt,
      ];
}

File: backend/models/SupportRequest.js
/**
 * Modèle Mongoose pour les demandes de support utilisateur.
 * Permet aux utilisateurs de signaler des bugs, poser des questions ou faire des suggestions.
 */

const mongoose = require('mongoose');

// --- Schéma principal ---

const supportRequestSchema = new mongoose.Schema({
    userId: {
        type: mongoose.Schema.Types.ObjectId,
        ref: 'User',
        required: true,
    },
    userEmail: {
        type: String,
        required: true,
    },
    userName: {
        type: String,
        required: true,
    },
    subject: {
        type: String,
        enum: ['bug', 'question', 'suggestion', 'other'],
        required: [true, 'Le sujet est requis'],
    },
    message: {
        type: String,
        required: [true, 'Le message est requis'],
        minlength: [10, 'Le message doit contenir au moins 10 caractères'],
        maxlength: [2000, 'Le message ne peut pas dépasser 2000 caractères'],
    },
    status: {
        type: String,
        enum: ['pending', 'in_progress', 'resolved', 'closed'],
        default: 'pending',
    },
    adminNotes: {
        type: String,
        default: null,
    },
    resolvedAt: {
        type: Date,
        default: null,
    },
    createdAt: {
        type: Date,
        default: Date.now,
    },
    updatedAt: {
        type: Date,
        default: Date.now,
    },
});

// --- Index ---

supportRequestSchema.index({ userId: 1, createdAt: -1 });

// Index pour l'admin (recherche par status)
supportRequestSchema.index({ status: 1, createdAt: -1 });

// --- Méthodes d'instance ---

/**
 * Retourne le libellé français du sujet de la demande.
 * @returns {string} Libellé du sujet
 */
supportRequestSchema.methods.getSubjectLabel = function () {
    const labels = {
        bug: 'Signalement de bug',
        question: 'Question',
        suggestion: 'Suggestion',
        other: 'Autre',
    };
    return labels[this.subject] || this.subject;
};

/**
 * Retourne le libellé français du statut de la demande.
 * @returns {string} Libellé du statut
 */
supportRequestSchema.methods.getStatusLabel = function () {
    const labels = {
        pending: 'En attente',
        in_progress: 'En cours',
        resolved: 'Résolu',
        closed: 'Fermé',
    };
    return labels[this.status] || this.status;
};

module.exports = mongoose.model('SupportRequest', supportRequestSchema);
